architecture/2024-10-user-defined-types-authoring-experience-feature-spec.md

```diff
@@ -0,0 +1,230 @@
+
+### Positive user outcome
+<!-- What is the positive outcome for the user if we deliver this, i.e. what is the value proposition for the user? Remember, this is user-centric. -->
+
+- Platform engineers can define the schema in a simple and intuitive way with minimal effort and learning curve.
+
+- Platform engineers can extend Radius with their custom services and provide self-service capabilities for application teams to use their custom services in Radius.
+
+- Platform engineers are empowered to enforce guardrails and best practices for application teams with the schema for the custom resource types.
+
+## Key scenarios
+<!-- List ~3-7 high level scenarios to clarify the value and point to how we will decompose this big area into component capabilities. We may ultimately have more than one level of scenario. -->
+The key scenarios for UDTs are outlined [here](/architecture/2024-06-resource-extensibility-feature-spec.md). We will decompose these scenarios specific to the authoring experience for defining the schema for the User-Defined-Type.
+
+Deb, a platform engineer at a Financial Services company who is responsible for building the custom internal developer platform. He has started the journey to enable radification one of their Budgets app. The Budgets app relies on an internal messaging service called Plaid. Deb wants to use Plaid as a resource type in Radius to deploy the application seamlessly and leverage all the features of Radius, such as Recipes, Connections, and the rad CLI, with ease and flexibility.
+
+### Scenario 1: Deb defines the schema for the internal messaging service Plaid
+<!-- One or two sentence summary -->
+As called out in the [feature spec](/architecture/2024-06-resource-extensibility-feature-spec.md), since `yaml` is the most common and standard format across open-source communities, we want to enable users to define the schema in a custom `yaml` format.
+
+1. Deb creates a sample schema definition for the internal messaging service Plaid
+   
+   Below are the basic set of inputs needed required to author a schema definition
+    
+    |Input | Description | Required|
+    |------|-------------|---------|
+    | Resource namespace | The namespace for the resource provider | Yes |
+    | Resource type name | The name of the resource type | Yes |
+    | Resource type description | The description of the resource type | No |
+    | Properties | The data model of the resource type | Yes |
+    | API Version | The version of the resource type | Yes |
+
+    Deb creates a plaid.yaml file with the following schema definition
+
+    ```yaml
+    namespace: 'Mycompany.Messaging'
+    resourceTypes:
+        'Mycompany.Messaging/plaidResource':
+            name: 'Name of the resource'
+            version: '2024-10-01'
```

Comment:
There are three *big* concepts for us to work with, and they are all nested.

- Resource provider
  - Resource type
    -  API version  

---

The simplest file structure is to define one resource provider per-file. Then you'd end up with a structure like:

```yaml
# Resource provider 
name: 'MyCompany.Messaging`
... # other resource provider settings go here

resourceTypes:
  # Resource type
  'plaidResource': 
     ... # other resource type settings go here

    apiVersions:
    # API version
      '2024-01-01': 
        schema: ...
        capabilities: ...
        ... # other api version settings go here
```

When you nest these concepts in YAML, they can either be map-like or array-like. I tend to prefer map-like because there's less ambiguity. (Your example is both, and it's redundant 😆).

Example map-like:

```yaml
# Resource provider 
resourceTypes:
  'plaidResource': 
    apiVersions:
      '2024-01-01': 
        schema: ...
        capabilities: ...
```

Example array-like:


```yaml
# Resource provider 
resourceTypes:
  - name: 'plaidResource'
    apiVersions:
      - name: '2024-01-01'
        schema: ...
        capabilities: ...
```

Reply:
Thanks for catching that. I have updated the format to be more map-like.